include/DDpackage.hpp

```diff
@@ -0,0 +1,2047 @@
+                }
+
+                if (!y.isTerminal() && y.p->v == w && e2.w != Complex::zero) {
+                    cn.returnToCache(e2.w);
+                }
+            }
+
+            auto e = makeDDNode(w, edge, true);
+            computeTable.insert({x.p, x.w}, {y.p, y.w}, {e.p, e.w});
+            return e;
+        }
+
+        ///
+        /// Matrix (conjugate) transpose
+        ///
+    public:
+        UnaryComputeTable<mEdge, mEdge> matrixTranspose{};
+        UnaryComputeTable<mEdge, mEdge> conjugateMatrixTranspose{};
+
+        mEdge transpose(const mEdge& a) {
+            if (a.p == nullptr || a.isTerminal() || a.p->symm) {
+                return a;
+            }
+
+            // check in compute table
+            auto r = matrixTranspose.lookup(a);
+            if (r.p != nullptr) {
+                return r;
+            }
+
+            std::array<mEdge, NEDGE> e{};
+            // transpose sub-matrices and rearrange as required
+            for (auto i = 0U; i < RADIX; ++i) {
+                for (auto j = 0U; j < RADIX; ++j) {
+                    e[RADIX * i + j] = transpose(a.p->e[RADIX * j + i]);
+                }
+            }
+            // create new top node
+            r = makeDDNode(a.p->v, e);
+            // adjust top weight
+            auto c = cn.getTemporary();
+            ComplexNumbers::mul(c, r.w, a.w);
+            r.w = cn.lookup(c);
+
+            // put in compute table
+            matrixTranspose.insert(a, r);
+            return r;
+        }
+        mEdge conjugateTranspose(const mEdge& a) {
+            if (a.p == nullptr)
+                return a;
+            if (a.isTerminal()) { // terminal case
+                auto r = a;
+                r.w    = ComplexNumbers::conj(a.w);
+                return r;
+            }
+
+            // check if in compute table
+            auto r = conjugateMatrixTranspose.lookup(a);
+            if (r.p != nullptr) {
+                return r;
+            }
+
+            std::array<mEdge, NEDGE> e{};
+            // conjugate transpose submatrices and rearrange as required
+            for (auto i = 0U; i < RADIX; ++i) {
+                for (auto j = 0U; j < RADIX; ++j) {
+                    e[RADIX * i + j] = conjugateTranspose(a.p->e[RADIX * j + i]);
+                }
+            }
+            // create new top node
+            r = makeDDNode(a.p->v, e);
+
+            auto c = cn.getTemporary();
+            // adjust top weight including conjugate
+            ComplexNumbers::mul(c, r.w, ComplexNumbers::conj(a.w));
+            r.w = cn.lookup(c);
+
+            // put it in the compute table
+            conjugateMatrixTranspose.insert(a, r);
+            return r;
+        }
+
+        ///
+        /// Multiplication
+        ///
+    public:
+        ComputeTable<mEdge, vEdge, vCachedEdge> matrixVectorMultiplication{};
+        ComputeTable<mEdge, mEdge, mCachedEdge> matrixMultiplication{};
```

Comment:
Should this variable be called `matrixMatrixMultiplication`?

Reply:
Probably 👍🏼